File: matiasmasso/Mat.NET/DTO/DTOOrderConfirmation.cs
using MatHelperStd;
using System;
using System.Collections.Generic;
using System.Linq;

namespace DTO
{
    public class DTOOrderConfirmation
    {
        public DTOEdiversaInterlocutor NADMS { get; set; } // message sender (us)
        public DTOEdiversaInterlocutor NADMR { get; set; } // message receiver (same as order message sender NadMr)
        public DTOEdiversaInterlocutor Supplier { get; set; } // NADSU
        public DTOEdiversaInterlocutor Buyer { get; set; } // NADBY
        public DTOEdiversaInterlocutor DeliveryPlatform { get; set; } // NADDP
        public DTOEdiversaInterlocutor InvoiceTo { get; set; } // NADDIV
        public DTOCur Currency { get; set; }
        public DTOPurchaseOrder Order { get; set; }
        public DateTime Fch { get; set; }
        public string Num { get; set; }
        public string OrderNumber { get; set; }
        public List<Item> Items { get; set; }
        public List<Exception> Exceptions { get; set; }

        public DTOOrderConfirmation() : base()
        {
            Items = new List<Item>();
            Exceptions = new List<Exception>();
        }

        public static DTOOrderConfirmation Factory(string sFilename, List<Exception> exs)
        {
            DTOOrderConfirmation retval = null;
            List<string> segments = TextHelper.ReadFileToStringList(sFilename, exs);
            if (segments != null)
                retval = Factory(segments);
            return retval;
        }

        public static DTOOrderConfirmation Factory(List<string> segments)
        {
            List<DTOEdiversaException> exs = new List<DTOEdiversaException>();
            DTOOrderConfirmation retval = new DTOOrderConfirmation();
            System.Globalization.CultureInfo provider = System.Globalization.CultureInfo.InvariantCulture;

            DTOOrderConfirmation.Item item = null;
            foreach (string segment in segments)
            {
                List<string> fields = segment.Split('|').ToList();
                switch (fields.First())
                {
                    case "BGM":
                        {
                            retval.Num = fields[1];
                            break;
                        }

                    case "DTM":
                        {
                            retval.Fch = DateTime.ParseExact(fields[1], "yyyyMMdd", provider);
                            break;
                        }

                    case "RFF":
                        {
                            if (fields.Count > 2 && fields[1] == "ON")
                                retval.OrderNumber = fields[2];
                            break;
                        }

                    case "NADSU":
                        {
                            retval.Supplier = new DTOEdiversaInterlocutor(segment);
                            break;
                        }

                    case "NADBY":
                        {
                            retval.Buyer = new DTOEdiversaInterlocutor(segment);
                            break;
                        }

                    case "NADDP":
                        {
                            retval.DeliveryPlatform = new DTOEdiversaInterlocutor(segment);
                            break;
                        }

                    case "NADIV":
                        {
                            retval.InvoiceTo = new DTOEdiversaInterlocutor(segment);
                            break;
                        }

                    case "CUX":
                        {
                            retval.Currency = DTOCur.Factory(fields[1]);
                            break;
                        }

                    case "LIN":
                        {
                            item = new DTOOrderConfirmation.Item(segment);
                            break;
                        }

                    case "PIALIN":
                        {
                            if (fields.Count > 2 && fields[1] == "SA" && item != null)
                                item.SupplierRef = fields[2];
                            break;
                        }

                    case "IMDLIN":
                        {
                            if (fields.Count > 3 && item != null)
                                item.Description = fields[3];
                            break;
                        }

                    case "QTYLIN":
                        {
                            if (fields.Count > 2 && item != null)
                                item.Qty = fields[2].toInteger();
                            break;
                        }

                    case "DTMLIN":
                        {
                            item.DeliveryTime = DateTime.ParseExact(fields[1], "yyyyMMdd", provider);
                            break;
                        }

                    case "PRILIN":
                        {
                            if (fields.Count > 2 && item != null)
                                item.Price = DTOEdiversaFile.ParseAmt(fields[2], exs).Eur;
                            break;
                        }
                }
            }
            return retval;
        }
        public class Item
        {
            public string Ean { get; set; }
            public int Lin { get; set; }
            public string SupplierRef { get; set; }
            public string Description { get; set; }
            public DTOProductSku Sku { get; set; }
            public int Qty { get; set; }
            public decimal Price { get; set; }
            public decimal Dto { get; set; }
            public DateTime DeliveryTime { get; set; }

            public Item(string segment) : base()
            {
                List<string> fields = segment.Split('|').ToList();
                if (fields.Count > 1)
                {
                    Ean = fields[1];
                    if (fields.Count > 3)
                        Lin = fields[3].toInteger();
                }
            }
        }
    }
}

File: matiasmasso/Mat.NET/DT2/DTOOrderConfirmation.vb
Public Class DTOOrderConfirmation
    Property Supplier As DTOEdiversaInterlocutor 'NADSU
    Property Buyer As DTOEdiversaInterlocutor 'NADBY
    Property DeliveryPlatform As DTOEdiversaInterlocutor 'NADDP
    Property InvoiceTo As DTOEdiversaInterlocutor 'NADDIV
    Property Currency As DTOCur
    Property Order As DTOPurchaseOrder
    Property Fch As Date
    Property Num As String
    Property OrderNumber As String
    Property Items As List(Of Item)
    Property Exceptions As List(Of Exception)

    Public Sub New()
        MyBase.New
        _Items = New List(Of Item)
        _Exceptions = New List(Of Exception)
    End Sub

    Shared Function Factory(sFilename As String, exs As List(Of Exception)) As DTOOrderConfirmation
        Dim retval As DTOOrderConfirmation = Nothing
        Dim segments As List(Of String) = TextHelper.ReadFileToStringList(sFilename, exs)
        If segments IsNot Nothing Then
            retval = Factory(segments)
        End If
        Return retval
    End Function

    Shared Function Factory(segments As List(Of String)) As DTOOrderConfirmation
        Dim retval As New DTOOrderConfirmation
        Dim provider As Globalization.CultureInfo = Globalization.CultureInfo.InvariantCulture

        Dim item As DTOOrderConfirmation.Item = Nothing
        For Each segment As String In segments
            Dim fields As List(Of String) = segment.Split("|").ToList
            Select Case fields.First
                Case "BGM"
                    retval.Num = fields(1)
                Case "DTM"
                    retval.Fch = Date.ParseExact(fields(1), "yyyyMMdd", provider)
                Case "RFF"
                    If fields.Count > 2 AndAlso fields(1) = "ON" Then
                        retval.OrderNumber = fields(2)
                    End If
                Case "NADSU"
                    retval.Supplier = New DTOEdiversaInterlocutor(segment)
                Case "NADBY"
                    retval.Buyer = New DTOEdiversaInterlocutor(segment)
                Case "NADDP"
                    retval.DeliveryPlatform = New DTOEdiversaInterlocutor(segment)
                Case "NADIV"
                    retval.InvoiceTo = New DTOEdiversaInterlocutor(segment)
                Case "CUX"
                    retval.Currency = DTOCur.Factory(fields(1))
                Case "LIN"
                    item = New DTOOrderConfirmation.Item(segment)
                Case "PIALIN"
                    If fields.Count > 2 AndAlso fields(1) = "SA" AndAlso item IsNot Nothing Then
                        item.SupplierRef = fields(2)
                    End If
                Case "IMDLIN"
                    If fields.Count > 3 AndAlso item IsNot Nothing Then
                        item.Description = fields(3)
                    End If
                Case "QTYLIN"
                    If fields.Count > 2 AndAlso item IsNot Nothing Then
                        item.Qty = fields(2)
                    End If
                Case "DTMLIN"
                    item.DeliveryTime = Date.ParseExact(fields(1), "yyyyMMdd", provider)
                Case "PRILIN"
                    If fields.Count > 2 AndAlso item IsNot Nothing Then
                        item.Price = fields(2)
                    End If
            End Select
        Next
        Return retval
    End Function
    Public Class Item
        Property Ean As String
        Property Lin As Integer
        Property SupplierRef As String
        Property Description As String
        Property Sku As DTOProductSku
        Property Qty As Integer
        Property Price As Decimal
        Property Dto As Decimal
        Property DeliveryTime As Date

        Public Sub New(segment As String)
            MyBase.New
            Dim fields As List(Of String) = segment.Split("|").ToList
            If fields.Count > 1 Then
                _Ean = fields(1)
                If fields.Count > 3 Then
                    _Lin = fields(3)
                End If
            End If
        End Sub
    End Class

End Class
